File: python/src/sum_best.py
# Return the shortest combination of any elements that add up to target_sum
from common import time_execution


@time_execution()
def best_sum_dp(target_sum, numbers):
    dp = [None] * (target_sum + 1)
    dp[0] = []  # Base case

    for i in range(target_sum + 1):
        if dp[i] is not None:
            for num in numbers:
                if i + num <= target_sum:
                    # If that entry doesn't exist yet or we found a shorter combination
                    if not dp[i+num] or len(dp[i])+1 < len(dp[i+num]):
                        dp[i+num] = [num]+dp[i]

    return dp[target_sum]


@time_execution()
def best_sum_rec(target_sum: int, numbers: list[int]) -> list[int]:
    def helper(target_sum: int) -> list[int]:
        if target_sum == 0:
            return []
        if target_sum < 0:
            return None

        shortest_combination = None

        for num in numbers:
            remainder = target_sum - num
            remainder_combination = helper(remainder)

            if remainder_combination is not None:
                combination = remainder_combination + [num]
                if shortest_combination is None or len(combination) < len(shortest_combination):
                    shortest_combination = combination

        return shortest_combination

    return helper(target_sum)


@time_execution()
def best_sum_rec_memo(target_sum: int, numbers: list[int]) -> list[int]:
    memo = {0: []}
    memo_count = 0

    def helper(target: int):
        nonlocal memo_count

        if target in memo:
            memo_count += 1
            return memo[target]
        if target < 0:
            return None

        shortest_combination = None

        for num in numbers:
            remainder = target-num

            if (remainder_combination := helper(remainder)) is not None:
                combination = remainder_combination + [num]

                if shortest_combination is None \
                        or len(combination) < len(shortest_combination):
                    shortest_combination = combination

        memo[target] = shortest_combination
        return memo[target]

    return f"{helper(target_sum)}, memo count: {memo_count}"


@time_execution()
def best_sum_rec_v2(target_sum: int, numbers: list[int]) -> list[int] | None:
    shortest_combination = None

    def helper(target: int, idx=0, nums_so_far=[]):
        nonlocal shortest_combination
        if target < 0 or idx == len(numbers):
            return None
        if target == 0:
            return nums_so_far

        # # Return early if we already have more elements in the combination
        if shortest_combination and len(nums_so_far) > len(shortest_combination):
            return None

        combination = helper(
            target - numbers[idx], idx, [numbers[idx]]+nums_so_far)
        if combination is not None:
            if shortest_combination is None or len(combination) < len(shortest_combination):
                shortest_combination = combination

        helper(target, idx + 1, nums_so_far)

    helper(target_sum)
    return shortest_combination


@time_execution()
def best_sum_rec_v2_memo(target_sum: int, numbers: list[int]) -> list[int] | None:
    memo = {0: []}
    memo_count = 0

    def helper(target: int, idx=0):
        nonlocal memo_count
        key = target

        if key in memo:
            memo_count += 1
            return memo[key]
        if target < 0 or idx == len(numbers):
            return None

        with_num = helper(target - numbers[idx], idx)
        if with_num is not None:
            with_num = with_num+[numbers[idx]]

        without_num = helper(target, idx + 1)

        if with_num is None or (without_num is not None and len(without_num) < len(with_num)):
            memo[key] = without_num
        else:
            memo[key] = with_num

        return memo[key]

    return f"{helper(target_sum)}, memo count: {memo_count}"


print(best_sum_dp(35, [1, 2, 3]))
# print(best_sum_rec(35, [1, 2, 3]))
print(best_sum_rec_memo(35, [1, 2, 3]))
print(best_sum_rec_v2(35, [1, 2, 3]))
print(best_sum_rec_v2_memo(35, [1, 2, 3]))

File: php/src/SumBest.php
<?php

function sum_best_dp(int $target, array $nums): array|null
{
    $dp = array_fill(0, $target + 1, null);
    $dp[0] = [];
    for ($i = 0; $i <= $target; $i++) {
        if ($dp[$i] !== null) {
            foreach ($nums as $num) {
                if ($i + $num <= $target) {
                    if ($dp[$i + $num] === null or (count($dp[$i]) + 1) < count($dp[$i + $num]))
                        $dp[$i + $num] = [$num, ...$dp[$i]];
                }
            }
        }
    }
    return $dp[$target];
}

function sum_best_rec(int $target, array $nums): array|null
{
    $helper = function (int $remainder) use ($nums, &$helper): array|null {
        if ($remainder < 0)
            return null;
        if ($remainder == 0)
            return [];

        $best = null;
        foreach ($nums as $num) {
            if (($res = $helper($remainder - $num)) !== null) {
                $combination = [...$res, $num];
                if ($best === null or count($combination) < count($best))
                    $best = $combination;
            }
        }

        return $best;
    };
    return $helper($target);
}


function sum_best_memo(int $target, array $nums): array|null
{
    $memo = [0 => []];
    $helper = function (int $remainder) use ($nums, &$memo, &$helper): array|null {
        if ($remainder < 0)
            return null;
        if (array_key_exists($remainder, $memo))
            return $memo[$remainder];

        $best = null;
        foreach ($nums as $num) {
            if (($res = $helper($remainder - $num)) !== null) {
                $combination = [...$res, $num];
                if ($best === null or count($combination) < count($best))
                    $best = $combination;
            }
        }
        $memo[$remainder] = $best;
        return $best;
    };
    return $helper($target);
}

function sum_best_rec_v2(int $target, array $nums): array|null
{
    $best = null;
    $helper = function (int $remainder, array $combination = [], int $idx = 0) use ($nums, &$best, &$helper) {
        if ($remainder < 0 or $idx == count($nums))
            return null;
        if ($remainder == 0)
            return $combination;
        $with = $helper($remainder - $nums[$idx], [$nums[$idx], ...$combination], $idx);
        if ($with !== null && ($best === null || count($with) < count($best))) {
            $best = $with;
        }

        $helper($remainder, $combination, $idx + 1);
    };

    $helper($target);
    return $best;
}

function sum_best_rec_v2_early(int $target, array $nums): array|null
{
    $best = null;
    $helper = function (int $remainder, array $combination = [], int $idx = 0) use ($nums, &$best, &$helper) {
        if ($remainder < 0 or $idx == count($nums))
            return null;
        if ($remainder == 0)
            return $combination;

        // Early return as soon as we go above best combination in length
        if ($best !== null and count($combination) > count($best))
            return null;

        $with = $helper($remainder - $nums[$idx], [$nums[$idx], ...$combination], $idx);
        if ($with !== null && ($best === null || count($with) < count($best))) {
            $best = $with;
        }

        $helper($remainder, $combination, $idx + 1);
    };

    $helper($target);
    return $best;
}

function sum_best_memo_v2(int $target, array $nums): array|null
{
    $memo = [0 => []];
    $helper = function (int $remainder, int $idx = 0) use ($nums, &$memo, &$helper): array|null {
        if ($remainder < 0 or $idx == count($nums))
            return null;
        if (array_key_exists($remainder, $memo))
            return $memo[$remainder];

        $with = $helper($remainder - $nums[$idx], 0);
        if ($with !== null)
            $with[] = $nums[$idx];

        $without = $helper($remainder, $idx + 1);

        // Pick the shortest combination so far
        if ($with === null or ($without !== null and count($without) < count($with)))
            $memo[$remainder] = $without;
        else
            $memo[$remainder] = $with;

        return $memo[$remainder];
    };

    return $helper($target);
}